renderdoc/driver/vulkan/vk_resources.h

```diff
@@ -1112,6 +1152,524 @@ typedef BitFlagIterator<VkImageAspectFlagBits, VkImageAspectFlags, int32_t> Imag
+    ValidateLayerRange(baseArrayLayer, layerCount, info.layerCount);
+    ValidateSliceRange(baseDepthSlice, sliceCount, info.extent.depth);
+  }
+};
+
+DECLARE_REFLECTION_STRUCT(ImageSubresourceRange);
+
+struct ImageSubresourceState
+{
+  uint32_t oldQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+  uint32_t newQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+  VkImageLayout oldLayout = UNKNOWN_PREV_IMG_LAYOUT;
+  VkImageLayout newLayout = UNKNOWN_PREV_IMG_LAYOUT;
+  FrameRefType refType = eFrameRef_None;
+
+  inline ImageSubresourceState() {}
+  inline ImageSubresourceState(const VkImageMemoryBarrier &barrier)
+      : oldQueueFamilyIndex(barrier.srcQueueFamilyIndex),
+        newQueueFamilyIndex(barrier.dstQueueFamilyIndex),
+        oldLayout(barrier.oldLayout),
+        newLayout(barrier.newLayout)
+  {
+  }
+  inline ImageSubresourceState(uint32_t queueFamilyIndex, VkImageLayout layout,
+                               FrameRefType refType = eFrameRef_None)
+      : oldQueueFamilyIndex(queueFamilyIndex),
+        newQueueFamilyIndex(queueFamilyIndex),
+        oldLayout(layout),
+        newLayout(layout),
+        refType(refType)
+  {
+  }
+  inline bool operator==(const ImageSubresourceState &other) const
+  {
+    return oldQueueFamilyIndex == other.oldQueueFamilyIndex &&
+           newQueueFamilyIndex == other.newQueueFamilyIndex && oldLayout == other.oldLayout &&
+           newLayout == other.newLayout && refType == other.refType;
+  }
+  inline bool operator!=(const ImageSubresourceState &other) const { return !(*this == other); }
+  void Update(const ImageSubresourceState &other, FrameRefCompFunc compose);
+  bool Update(const ImageSubresourceState &other, ImageSubresourceState &result,
+              FrameRefCompFunc compose) const;
+};
+
+DECLARE_REFLECTION_STRUCT(ImageSubresourceState);
+
+struct TaggedImageSubresourceState
+{
+  ImageSubresourceRange range;
+  ImageSubresourceState state;
+};
+
+DECLARE_REFLECTION_STRUCT(TaggedImageSubresourceState);
+
+struct ImgRefs;
+
+class ImageSubresourceMap
+{
+  friend class SubresourceRangeIterator;
+
+  rdcarray<ImageSubresourceState> m_values;
+  VkImageAspectFlags m_aspectMask = 0u;
+  ImageInfo m_imageInfo;
+  enum class FlagBits : uint16_t
+  {
+    AreAspectsSplit = 0x1,
+    AreLevelsSplit = 0x2,
+    AreLayersSplit = 0x4,
+    IsDepthSplit = 0x8,
+    IsUninitialized = 0x8000,
+  };
+  uint16_t m_flags = 0;
+  uint16_t m_aspectCount = 0;
+  inline static bool AreAspectsSplit(uint16_t flags)
+  {
+    return (flags & (uint16_t)FlagBits::AreAspectsSplit) != 0;
+  }
+  inline static bool AreLevelsSplit(uint16_t flags)
+  {
+    return (flags & (uint16_t)FlagBits::AreLevelsSplit) != 0;
+  }
+  inline static bool AreLayersSplit(uint16_t flags)
+  {
+    return (flags & (uint16_t)FlagBits::AreLayersSplit) != 0;
+  }
+  inline static bool IsDepthSplit(uint16_t flags)
+  {
+    return (flags & (uint16_t)FlagBits::IsDepthSplit) != 0;
+  }
```

Comment:
Depth splitting is essentially the same behaviour that was previously implemented in `ImgRefs`. This allows us to track the read/write state independently for each depth slices.

It should not be possible for subresources which differ only in depth to ever have different layouts. I will make sure this is handled correctly for automatically applied layout transitions (TODO).

When the image state is used to generate barriers, subresources with non-zero `baseDepthSlice` are skipped, so a single barrier is generated for the entire depth range, whether or not the state is split by depth.

The motivation here is that finer grain read/write state should improve the reset performance. Even if we don't try to apply image initial contents at the depth slice level, there will still be a benefit in certain cases. E.g., if the application reads from depth slice 0, and then writes to depth slice 1, that does not require a reset. But if we drop the depth slice info, that becomes a read-before-write, which does require a reset.

Reply:
Ah right I see, so it's split up for the benefit of reference tracking rather than for layout tracking.

I'm a little skeptical that there would be enough gain for this to be worth the added complexity, but admittedly I have not nearly as much profiling of initial states application as you have. Is this a scenario you've seen come up for real workloads? If that's the case, or if you have high confidence it's going to have a measurable benefit, I'm fine with this going in. I primarily raised the issue because I thought it was being added for correctness (since this is an odd quirk of the 3D/2D aliasing behaviour).